frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -642,13 +680,41 @@ export default ({
           messageId: null
         })
       }
-    }, 500)
-  },
-  watch: {
-    currentChatRoomId (to, from) {
+    }, 500),
+    archiveMessageState (chatRoomId = null) {
+      if (!this.isJoinedChatRoom(chatRoomId)) {
+        return
+      }
+      const unit = this.chatRoomSettings?.actionsPerPage || CHATROOM_ACTIONS_PER_PAGE
+      const from = this.latestEvents.length ? GIMessage.deserialize(this.latestEvents[0]).hash() : null
+      const to = this.latestEvents.length
+        ? GIMessage.deserialize(this.latestEvents[this.latestEvents.length - 1]).hash()
+        : null
+
+      // NOTE: save messages in the browser storage, but not more than 3 pages of events
+      if (this.latestEvents.length >= 3 * unit) {
+        sbp('gi.db/archive/delete', this.getArchiveKeyFromChatRoomId(chatRoomId))
+      } else if (to !== this.messageState.prevTo || from !== this.messageState.prevFrom) {
+        // this.currentChatRoomId could be wrong when the channels are switched very fast
+        // so it's good to initiate using input parameter chatRoomId
+        sbp('gi.db/archive/save', this.getArchiveKeyFromChatRoomId(chatRoomId), JSON.stringify(this.latestEvents))
+      }
+    },
+    getArchiveKeyFromChatRoomId (chatRoomId) {
+      const curChatRoomId = chatRoomId || this.currentChatRoomId
+      return `messages/${curChatRoomId}`
+    },
+    refreshContent: debounce(function (from, to) {
```

Comment:
Below this I see a call to `setMessageEventListener`, which seems to set a lot of event listeners. But I don't see a corresponding `'okTurtles.events/off'` call for all of these:

```js
      if (force) {
        sbp('okTurtles.events/on', `${CHATROOM_MESSAGE_ACTION}-${to || this.currentChatRoomId}`, this.listenChatRoomActions)
      } else if (this.isJoinedChatRoom(to)) {
        sbp('okTurtles.events/on', `${CHATROOM_MESSAGE_ACTION}-${to}`, this.listenChatRoomActions)
      }
```

This seems like something I didn't catch in a previous PR review.

Reply:
I think we remove the event listeners in these two places by calling `sbp('okTurtles.events/off')`.
1. In `beforeDestroy` life cycle
2. in `setMessageEventListener` function

Reply:
Hmm, you're right. I didn't understand that `setMessageEventListener` properly handled this